Integrate ALE with Neovim LSP Client & Lua

w0rp opened the pull request:
This pull request is a major move toward goals stated in #4815. The changes here accomplish the following.

1. ALE will use Neovim's LSP client in Neovim 0.8 and up by default.
2. ALE can be configured via `ale.setup` and `ale.setup.buffer` functions in Lua.
3. All ALE settings have documentation updated to describe how to configure ALE easily either in Vim script or in Lua.
4. Various ALE settings have been tweaked so Boolean values work equally well as Numbers, to make configuring ALE easier for Neovim users.
5. Various basic ALE functions have been mirrored in one way or another in Lua for easier configuration in Lua.
6. Cover all new and previous Lua code with tests with [busted](https://lunarmodules.github.io/busted/).

This is a major leap forward for integrating ALE much better with built-in Neovim tools, Neovim plugins, and for making it easy to configure ALE in Neovim user configurations that are purely written in Lua.

w0rp committed:
Start up language servers with Neovim's API

Get language servers starting and displaying diagnostics with Neovim's
API in Neovim 0.8 and up. With this set up, now ALE needs to take over
handling diagnostics returned by the language servers.

w0rp committed:
Handle Neovim LSP diagnostics via ALE's functions

w0rp committed:
Remove the log_level line from Lua code

w0rp committed:
Handle other LSP capabilities via ALE

Save capabilities from language servers ALE connects to via Neovim's LSP
client and handle responses to requests ALE sends to language servers.
This enables ALE to work with Neovim's language client for its commands
while also letting users directly use the connected clients for native
Neovim keybinds and so on.

w0rp committed:
Clean up message handling for LSP integration

Now we ought to be able to handle any kind of response for any request
we send, clean up message handling so there are fewer changes to make
to LSP code to adjust for Neovim integration.

w0rp committed:
Update documentation for Neovim LSP integration

Update documentation to advertise ALE's integration with Neovim's native
LSP client, and explain how functionality is integrated with ALE,
Neovim's native tools, and other plugins.

w0rp committed:
Fix tests for Neovim LSP integration

w0rp committed:
Supply language_id values to Neovim LSP API

Change logic so ALE's LSP implementation and the Neovim LSP client
retrieve the language_id for language clients at roughly the same time
via the same means. This makes ALE inform the language server what the
language for the language is for clients.

w0rp committed:
Add Neovim TCP connections to language servers

Support TCP connections to language servers through Neovim's built in
client. In all but what is currently the nightly builds of Neovim
connections via a hostname will fail, but connections via an IP address
should function. We will still enable the built in Neovim client by
default anyway, as LSP clients very rarely connect over TCP.

w0rp committed:
Support g:ale_shell with the Neovim LSP client

w0rp committed:
Implement Lua ALE setup & overhaul documentation

1. Add ale.setup and ale.setup.buffer for pure Lua configuration.
2. Update many global settings to use Booleans instead of numbers to
   make types easiert to work with in Lua.
3. Radically reformat documentation and fix errors to make
   documentation more usable for Neovim users.

w0rp committed:
Detect the Lua project root using .luarc.json

w0rp committed:
Add basic Lua ALE functions and test coverage

Ensure that basic ALE functions `ale.var`, `ale.escape`, and `ale.env`
are available in Lua. Cover all Lua code so far with busted tests,
fixing bugs where ALE variables can be set with Boolean values instead
of numbers. Document all functionality so far.

w0rp committed:
Add ale.has to mirror ale#Has in Lua

w0rp committed:
Implement ale.get_filename_mappings in Lua

w0rp committed:
Implement ale.pad in Lua

w0rp committed:
Implement ale.queue for calling ale#Queue in Lua

w0rp commented:
@hsanson Check this one out with your configuration and let me know if you run into any issues. I took a career break for just over a month from my day job to go on a holiday and then work on this.

hsanson commented:
@w0rp I converted my configuration to use setup() method and seems to work fine so far.   I will be using this new configuration next week at work to stress test it.

Maybe a note in the documentation saying that the `ale_` prefix of the configuration options is not needed would be good.  Initially I had something like below that did not work and then I re-read the help and realized the `ale_` prefixes are not needed:

```lua
    require("ale").setup({
      ale_disable_lsp = "auto",
      ale_sign_info = g.symbol_info,
      ale_sign_error = g.symbol_error,
      ale_sign_warning = g.symbol_warn,
      ale_sign_priority = 4,
      ale_open_list = 0,
      ale_echo_msg_format = "%severity% [%linter%] (%code%) - %s",
       .....
  })
```

w0rp commented:
Awesome, thanks. I'm going to leave this branch open for a bit while you use it for a week or so before merging it, so I don't suddenly break something important for users. I'll prepare another branch I'll rebase later for some additional changes for the goals in that issue.

If you can suggest some edits somewhere in documentation to make the settings names clearer, let me know, or feel free to push a commit. I updated the documention for every option I could see to mention the them under both names, and there's examples in a number of places explaining how to use `ale.setup` and `ale.setup.buffer`.

w0rp committed:
Close #3600 - Implement pull diagnostics in VimL

Implement pull diagnostics in the VimL implementation so ALE is able
to track when servers are busy checking files. Only servers that
support this feature will return diagnostics these ways.

w0rp committed:
#3600 Implement pull model with Neovim Client

Implement the diagnostics pull model with the LSP Neovim client. We
must handle messages a little different and tweak client capabilities
for pull diagnostics to work through the Neovim client.

hsanson commented:
@w0rp I have been using  lua configurtion without any issues at home and work for few days already. I believe there are no issues and this can be merged.

w0rp commented:
@hsanson Nice. Thanks for testing this. I'll merge this now and the other PR. I'm sure someone somewhere will have some issues, but we'll discover those and fix them as issue come in.
